private/pkg/manifest/manifest.go

```diff
@@ -0,0 +1,269 @@
+}
+
+func NewDigestFromString(typedDigest string) (*Digest, error) {
+	hashfunc, digestStr, found := strings.Cut(typedDigest, ":")
+	if !found {
+		return nil, errors.New("malformed digest string")
+	}
+	return NewDigestFromHex(hashfunc, digestStr)
+}
+
+func (d *Digest) String() string {
+	return d.Type() + ":" + d.Hex()
+}
+
+func (d *Digest) Type() string {
+	return d.dtype
+}
+
+func (d *Digest) Bytes() []byte {
+	return d.digest
+}
+
+func (d *Digest) Hex() string {
+	return d.hexstr
+}
+
+// ManifestError occurs when a manifest is malformed.
+type ManifestError struct {
+	lineno int
+	msg    string
+}
+
+var _ error = (*ManifestError)(nil)
+
+func newManifestError(lineno int, msg string) *ManifestError {
+	return &ManifestError{
+		lineno: lineno,
+		msg:    msg,
+	}
+}
+
+func (e *ManifestError) Error() string {
+	return fmt.Sprintf("invalid manifest: %d: %s", e.lineno, e.msg)
+}
+
+// Manifest represents a list of paths and their digests.
+type Manifest struct {
+	paths   map[string]*Digest
+	digests map[string]string
+	hash    sha3.ShakeHash
+}
+
+var _ encoding.TextMarshaler = (*Manifest)(nil)
+var _ encoding.TextUnmarshaler = (*Manifest)(nil)
+
+// NewManifest creates an empty manifest.
+func NewManifest() *Manifest {
+	return &Manifest{
+		paths:   make(map[string]*Digest),
+		digests: make(map[string]string),
+		hash:    sha3.NewShake256(),
+	}
+}
+
+func splitManifest(data []byte, atEOF bool) (int, []byte, error) {
+	// Return a line without LF.
+	if i := bytes.IndexByte(data, '\n'); i >= 0 {
+		return i + 1, data[0:i], nil
+	}
+
+	// EOF occurred with a partial line.
+	if atEOF && len(data) != 0 {
+		return 0, nil, errPartial
+	}
+
+	return 0, nil, nil
+}
+
+// NewManifestFromReader builds a manifest an encoded manifest reader.
+func NewManifestFromReader(manifest io.Reader) (*Manifest, error) {
+	m := NewManifest()
+	scanner := bufio.NewScanner(manifest)
+	scanner.Split(splitManifest)
+	lineno := 0
+	for scanner.Scan() {
+		lineno++
+		hash, path, found := strings.Cut(scanner.Text(), "  ")
+		if !found {
+			return nil, newManifestError(lineno, "invalid entry")
+		}
+		digest, err := NewDigestFromString(hash)
+		if err != nil {
+			return nil, multierr.Append(
+				newManifestError(lineno, "invalid hash"),
+				err,
+			)
+		}
+		if digest.Type() != shake256Name {
+			return nil, newManifestError(lineno, "unknown hash")
+		}
+		if len(digest.Bytes()) != 64 {
+			return nil, newManifestError(lineno, "short digest")
+		}
+		m.addDigest(path, digest)
+	}
+	err := scanner.Err()
+	if err == errPartial {
+		return nil, newManifestError(lineno, "partial record")
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	return m, nil
+}
+
+// NewManifestFromBucket creates a manifest from a bucket.
+func NewManifestFromBucket(
+	ctx context.Context,
+	bucket storage.ReadBucket,
+) (*Manifest, error) {
+	m := NewManifest()
+	err := bucket.Walk(ctx, "", func(info storage.ObjectInfo) error {
+		path := info.Path()
+		obj, err := bucket.Get(ctx, path)
```

Comment:
And related, curious how is useful to get a Manifest from a whole bucket?

Reply:
> If you're passing a storagemem in here, don't think you can do .Get inside of a .Walk, will cause race issues. You'll need to first grab all the paths in .Walk, and then loop to get each content.

Wow. That seems less than ideal. Maybe I'll fix that because that's not a constraint that's documented and what I think I'm doing isn't crazy. It's pretty common to list a directory and then do things like stat and read them. Walk + Get probably shouldn't interfere with each other.

> And related, curious how is useful to get a Manifest from a whole bucket?

Buckets are created for modules, for the proto files, config, docs, and license files. This construction is to be used when pushing as the push can build a manifest for a module bucket. That said, this will probably move the storage package.

Reply:
Yeah, they both use the same `RLock` ([Walk](https://github.com/bufbuild/buf/blob/f1becc1a99c5f8c8b1ea8013f9aba83885a16195/private/pkg/storage/storagemem/bucket.go#L61), [Get](https://github.com/bufbuild/buf/blob/f1becc1a99c5f8c8b1ea8013f9aba83885a16195/private/pkg/storage/storagemem/bucket.go#L44))

Reply:
I think while the `RLock` is there, you should first get the list of paths, and then outside of the `Walk` fetch its contents.

Reply:
Turns out the RLock is fine. `Walk`, `Get`, and `Stat` takes read locks and `Put`, `Delete`, and `DeleteAll` take write locks. So it's fine to call `Get` in `Walk`'s callback. There is an opportunity to make `Put` and others work too, though I'll defer that work for now.